x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/DateTimeArithmeticOperation.java

```diff
@@ -55,31 +79,100 @@ ExpressionEvaluator apply(
     protected TypeResolution resolveType() {
         DataType leftType = left().dataType();
         DataType rightType = right().dataType();
-        // date math is only possible if one argument is a DATETIME and the other a (foldable) TemporalValue
+
+        // Date math is only possible if either
+        // - one argument is a DATETIME and the other a (foldable) TemporalValue, or
+        // - both arguments are TemporalValues (so we can fold them).
         if (isDateTimeOrTemporal(leftType) || isDateTimeOrTemporal(rightType)) {
-            if (argumentOfType(DataTypes::isDateTime) == null || argumentOfType(EsqlDataTypes::isTemporalAmount) == null) {
-                return new TypeResolution(
-                    format(null, "[{}] has arguments with incompatible types [{}] and [{}]", symbol(), leftType, rightType)
-                );
+            if ((leftType == DataTypes.DATETIME && isTemporalAmount(rightType))
+                || (rightType == DataTypes.DATETIME && isTemporalAmount(leftType))) {
+                return TypeResolution.TYPE_RESOLVED;
+            }
+            if (leftType == TIME_DURATION && rightType == TIME_DURATION) {
+                return TypeResolution.TYPE_RESOLVED;
+            }
+            if (leftType == DATE_PERIOD && rightType == DATE_PERIOD) {
+                return TypeResolution.TYPE_RESOLVED;
             }
-            return TypeResolution.TYPE_RESOLVED;
+
+            return new TypeResolution(
+                format(null, "[{}] has arguments with incompatible types [{}] and [{}]", symbol(), leftType, rightType)
+            );
         }
         return super.resolveType();
     }
 
+    /**
+     * Override this to allow processing literals of type {@link EsqlDataTypes#DATE_PERIOD} when folding constants.
+     * Used in {@link DateTimeArithmeticOperation#fold}.
+     * @param left the left period
+     * @param right the right period
+     * @return the result of the evaluation
+     */
+    abstract Period computeDateArithmetic(Period left, Period right);
```

Comment:
nit: could be just `fold(Period, Period)` (and corresponding `fold(Duration, Duration` below) to avoid the "full" date confusion.